test_data/proxy/src/com/intellij/memory/agent/MemoryAgent.java

```diff
@@ -0,0 +1,207 @@
+package com.intellij.memory.agent;
+
+import java.io.File;
+import java.lang.reflect.Array;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+/**
+ * <p>Memory agent allows to perform heap diagnostics or attach allocation listeners for
+ * <a href="https://openjdk.java.net/jeps/331">low-overhead heap profiling</a> using
+ * <a href="https://docs.oracle.com/javase/8/docs/technotes/guides/jvmti/">JVMTI</a>.<p/>
+ *
+ * <p>To access the memory agent's instance use {@link #get()}. If the binary wasn't loaded
+ * successfully you can get the corresponding exception by calling {@link #getLoadingException()}.<p/>
+ *
+ * <p>All methods of this class (except for {@link #get()}) throw {@link com.intellij.memory.agent.MemoryAgentExecutionException}
+ * if a call to native method failed.<p/>
+ */
+public class MemoryAgent {
+    private static final String allocationSamplingIsNotSupportedMessage = "Allocation sampling is not supported for this version of jdk";
+    private static final String agentLibraryWasNotLoadedMessage = "Agent library wasn't loaded";
+    private static final String failedToCallNativeMethodMessage = "Failed to call native method";
+
+    private final IdeaNativeAgentProxy proxy;
+    private final List<AllocationListenerInfo> listeners;
```

Comment:
Can't we have the only array with listener instead two? Having copy-on-write array with listener here and reading it from the native side might work. WDYT?